src/NUnitFramework/framework/Constraints/Comparers/ComparisonState.cs

```diff
@@ -0,0 +1,59 @@
+using System;
+using NUnit.Framework.Internal;
+
+namespace NUnit.Framework.Constraints.Comparers
+{
+    internal struct ComparisonState
+    {
+        /// <summary>
+        /// Flag indicating whether or not this is the top level comparison.
+        /// </summary>
+        public readonly bool TopLevelComparison;
+
+        /// <summary>
+        /// A list of tracked comparisons
+        /// </summary>
+        private readonly ImmutableStack<Comparison> _comparisons;
+
+        public ComparisonState(bool topLevelComparison)
+            : this(topLevelComparison, ImmutableStack<Comparison>.Empty)
+        {
+        }
+
+        private ComparisonState(bool topLevelComparison, ImmutableStack<Comparison> comparisons)
+        {
+            TopLevelComparison = topLevelComparison;
+            _comparisons = comparisons;
+        }
+
+        public ComparisonState PushComparison(object x, object y)
+        {
+            return new ComparisonState(
+                false,
+                _comparisons.Push(new Comparison(x, y))
+            );
+        }
+
+        public bool DidCompare(object x, object y)
+        {
+            foreach (var comparison in _comparisons)
+                if (Object.ReferenceEquals(comparison.X, x) &&
+                    Object.ReferenceEquals(comparison.Y, y))
```

Comment:
Keywords for built-in types are recommended over type names. If you prefer, `ReferenceEquals` doesn't need to be qualified with `Object.` or `object.` because all types inherit from object.